MotherOfAllQueries.sql

```diff
@@ -0,0 +1,407 @@
+		LEFT JOIN clients on clients.SocialGroupId = socialGroups.Id
+		LEFT JOIN accounts on accounts.ClientId = clients.Id
+		LEFT JOIN cards on cards.AccountId = accounts.Id
+	WHERE socialGroups.Id = socGroupsHigh.Id) AS CardsCount
+FROM socialGroups as socGroupsHigh
+GO
+
+-- Task 6 
+CREATE PROCEDURE Add10$ToSocGroup 
+	@SocGroupId INT
+AS
+BEGIN
+	IF @SocGroupId < 1
+		BEGIN
+			PRINT 'Invalid social group ID'
+		END;
+	ELSE
+		IF NOT EXISTS (SELECT * 
+			FROM socialGroups
+			WHERE socialGroups.Id = @SocGroupId)
+			BEGIN
+				PRINT 'Social group with this ID does not exist'
+			END;
+		ELSE
+		IF NOT EXISTS (SELECT *
+			FROM clients
+			WHERE clients.SocialGroupId = @SocGroupId)
+			BEGIN
+				PRINT 'No clients of that social group found'
+			END;
+			ELSE
+				IF NOT EXISTS (SELECT *
+					FROM accounts
+						JOIN clients on accounts.ClientId = clients.Id
+					WHERE clients.SocialGroupId = @SocGroupId)
+					BEGIN
+						PRINT 'No accounts associated with clients of that social group found'
+					END
+				ELSE
+				BEGIN
+					UPDATE accounts
+					SET Balance += 10
+					FROM accounts
+						JOIN clients on clients.Id = accounts.ClientId
+					WHERE clients.SocialGroupId = @SocGroupId
+				END;
+END;
+GO
+
+SELECT accounts.Balance
+FROM accounts
+
+EXEC Add10$ToSocGroup 1
+
+SELECT accounts.Balance
+FROM accounts
+GO
+
+-- Task 7
+SELECT accounts.Id, clients.Name, clients.Surname, accounts.Balance - COALESCE(SUM(cards.Balance),0) AS AvailableFunds
+FROM accounts
+	JOIN clients on accounts.ClientId = clients.Id
+	LEFT JOIN cards on cards.AccountId = accounts.Id
+GROUP BY accounts.Id, clients.Name, clients.Surname, accounts.Balance
+GO
+
+-- Task 8
+CREATE PROCEDURE TransferMoneyToCardFromAcc 
+	@AccountId INT,
+	@CardId INT,
+	@Amount MONEY
+AS
+BEGIN
+	BEGIN TRY
+	BEGIN TRANSACTION
+	IF @AccountId < 1 OR @CardId < 1 OR @Amount<=0
+		RAISERROR ('Invalid input', 16, 1);
+	IF NOT EXISTS (SELECT * 
+		FROM cards
+		WHERE cards.Id = @CardId
+			AND cards.AccountId = @AccountId)
+		RAISERROR ('Invalid ID or IDs', 16, 1);
+	IF NOT EXISTS (SELECT accounts.Id, accounts.Balance
+		FROM accounts
+			JOIN cards on cards.AccountId = accounts.Id
+		WHERE accounts.Id = @AccountId
+		GROUP BY accounts.Id, accounts.Balance
+		HAVING accounts.Balance - SUM(cards.Balance) >= @Amount)
+		RAISERROR ('Not enough money on account to transfer', 16, 1);
+	UPDATE cards
+	SET Balance += @Amount
+	FROM cards
+	WHERE Id = @CardId
+	COMMIT
+	END TRY
+	BEGIN CATCH
+		DECLARE @ErrorMessage NVARCHAR(4000);
+		DECLARE @ErrorSeverity INT;
+		DECLARE @ErrorState INT;
+
+		SET @ErrorMessage = ERROR_MESSAGE();
+        SET @ErrorSeverity = ERROR_SEVERITY();
+        SET @ErrorState = ERROR_STATE();
+		RAISERROR (@ErrorMessage, @ErrorSeverity, @ErrorState);
+		ROLLBACK
+	END CATCH
+END;
+GO
+
+SELECT cards.Id, cards.Balance
+FROM cards
+
+EXEC TransferMoneyToCardFromAcc 1, 1, -100
+
+SELECT cards.Id, cards.Balance
+FROM cards
+GO
+
+--Task 9
+CREATE TRIGGER Accounts_UPDATE
```

Comment:
![image](https://github.com/Sailtor/Innowise.SQLTestTask/assets/96731963/b56544aa-9f3f-4022-ac0c-ef80d2298df0)

Please review your script and fix the next error. Also there woud be problems if you would try to update other fields (not balance). So better to choose another type of trigger here or to take into considerations all of this

Reply:
The same applies to the next trigger